File: ruby/median-of-two-sorted-arrays.rb
#!/usr/bin/env ruby

# There are two sorted arrays nums1 and nums2 of size m and n respectively.

# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

# Example 1:
# nums1 = [1, 3]
# nums2 = [2]

# The median is 2.0
# Example 2:
# nums1 = [1, 2]
# nums2 = [3, 4]

# The median is (2 + 3)/2 = 2.5

# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Float}
def find_median_sorted_arrays(nums1, nums2)
  nums1, nums2 = nums2, nums1 if nums1.size > nums2.size 
  if nums1[-1] <= nums2[0]
    return nums1.size == nums2.size ? median_of_array([nums1[-1], nums2[0]]) : median_of_array(nums2[0...-(nums1.size)]) 
  elsif nums1[0] >= nums2[-1]
    return nums1.size == nums2.size ? median_of_array([nums1[0], nums2[-1]]) : median_of_array(nums2[nums1.size..-1]) if nums1[0] >= nums2[-1]
  else
    median1 = median_of_array(nums1)
    median2 = median_of_array(nums2)
    if median1 == median2
      return median1
    elsif median1 < median2      
      return find_median_sorted_arrays()
    else
      
    end
  end
    
end


def half_of_array(nums, left)
  return nil if nums.empty?
  return nums[0] if nums.size == 1
  div, mod = nums.size.divmod 2

end

def median_of_array(nums)
  return nil if nums.empty?
  return nums[0] if nums.size == 1
  div, mod = nums.size.divmod 2
  if mod == 0
    (nums[div] + nums[div - 1]) / 2.0
  else
    nums[div]
  end
end

File: swift/t_code/Source/Hard/MedianOfTwoSortedArrays.swift
//
//  MedianOfTwoSortedArrays.swift
//  t_code
//
//  Created by Tong on 07/07/2018.
//  Copyright © 2018 LuckyTR. All rights reserved.
//

import Foundation

/*
 
 There are two sorted arrays nums1 and nums2 of size m and n respectively.
 
 Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
 
 Example 1:
 nums1 = [1, 3]
 nums2 = [2]
 
 The median is 2.0
 Example 2:
 nums1 = [1, 2]
 nums2 = [3, 4]
 
 The median is (2 + 3)/2 = 2.5

 
 */

class MedianOfTwoSortedArrays {
    class Solution {
        func findMedianSortedArrays(_ nums1: [Int], _ nums2: [Int]) -> Double {
            let totalCount = nums1.count + nums2.count
            var nums = [Int](repeating: 0, count: totalCount)
            
            
            var i = 0, j = 0, k = 0
            
            while i < nums1.count && j < nums2.count {
                if nums1[i] <= nums2[j] {
                    nums[k] = nums1[i]
                    i += 1
                } else {
                    nums[k] = nums2[j]
                    j += 1
                }
                k += 1
            }
            
            if i < nums1.count {
                nums.replaceSubrange(k..<totalCount, with: nums1[i..<nums1.count])
            }
            
            if j < nums2.count {
                nums.replaceSubrange(k..<totalCount, with: nums2[j..<nums2.count])
            }
            
            return medianFromArray(nums)
        }
        
        func medianFromArray(_ array:[Int]) -> Double {
            guard !array.isEmpty else {
                return 0
            }
            
            if array.count % 2 == 0 {
                let idx = array.count / 2
                return Double(array[idx] + array[idx - 1]) / 2.0
            } else {
                return Double(array[array.count / 2])
            }
        }
    }
}


